spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/RemoveNetCarryoverIterator.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark;
+
+import java.util.Iterator;
+import org.apache.iceberg.MetadataColumns;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.types.StructType;
+
+/**
+ * This class computes the net changes across multiple snapshots. It takes a row iterator, and
+ * assumes the following:
+ *
+ * <ul>
+ *   <li>The row iterator is partitioned by all columns.
+ *   <li>The row iterator is sorted by all columns, change order, and change type. The change order
+ *       is 1-to-1 mapping to snapshot id.
+ * </ul>
+ */
+public class RemoveNetCarryoverIterator extends RemoveCarryoverIterator {
+
+  private final int[] indicesToIdentifySameRow;
+
+  private Row cachedNextRow = null;
+  private Row cachedRow = null;
+  private long cachedRowCount = 0;
+
+  protected RemoveNetCarryoverIterator(Iterator<Row> rowIterator, StructType rowType) {
+    super(rowIterator, rowType);
+    this.indicesToIdentifySameRow = generateIndicesToIdentifySameRow();
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (cachedRowCount > 0) {
+      return true;
+    }
+
+    if (cachedNextRow != null) {
+      return true;
+    }
+
+    return rowIterator().hasNext();
+  }
+
+  @Override
+  public Row next() {
+    // if there are cached rows, return one of them from the beginning
+    if (cachedRowCount > 0) {
+      cachedRowCount--;
+      return cachedRow;
+    }
+
+    Row currentRow = getCurrentRow();
+
+    // return it directly if the current row is the last row
+    if (!rowIterator().hasNext()) {
+      return currentRow;
+    }
+
+    Row nextRow = rowIterator().next();
```

Comment:
When `cachedRowCount` is larger than zero, we have to drain it first, then get the cached next row if there is one.

Reply:
```
(1, 'a', insert)
(1, 'a', insert)
(1, 'b', insert)
(1, 'b', delete)
```
Here is an example, we need to drain 'a' first, then get the cached next row (1, 'b', insert), and try to match it with the row (1, 'b', delete)

Reply:
Maybe not clear:

```

  @Override
  public Row next() {
    // if there are cached rows, return one of them from the beginning
    if (cachedRowCount > 0) {
      cachedRowCount--;
      return cachedRow;
    }

    this.cachedRow = getCurrentRow();

    // return it directly if the current row is the last row
    if (!rowIterator().hasNext()) {
      return cachedRow;
    }

    this.cachedNextRow = rowIterator().next();
    cachedRowCount = 1;

    // pull rows from the iterator until two consecutive rows are different
    while (isSameRecord(cachedRow, cachedNextRow)) {
      if (oppositeChangeType(cachedRow, cachedNextRow)) {
        // two rows with opposite change types means no net changes, remove both
        cachedRowCount--;
      } else {
        // two rows with same change types means potential net changes, cache the next row, reset it
        // to null
        cachedRowCount++;
      }

      // stop pulling rows if there is no more rows or the next row is different
      if (cachedRowCount <= 0 || !rowIterator().hasNext()) {
        this.cachedNextRow = null;
        break;
      }

      this.cachedNextRow = rowIterator().next();
    }

    return null;
  }
```

I think it will work to remove 'currentRow'.  But I am not sure if I am too aggressive for removing 'nextRow'.  Please check if I made a mistake.

Reply:
Made the change accordingly. Thanks for the suggestion